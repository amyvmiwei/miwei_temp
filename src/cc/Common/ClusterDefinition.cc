/*
 * Copyright (C) 2007-2015 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 3 of the
 * License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

/// @file
/// Definitions for ClusterDefinition.
/// This file contains type definitions for ClusterDefinition, a class that
/// caches and provides access to the information in a cluster definition file.

#include <Common/Compat.h>

#include "ClusterDefinition.h"

#include <Common/HostSpecification.h>
#include <Common/Logger.h>
#include <Common/ClusterDefinition/Tokenizer.h>

#include <boost/algorithm/string.hpp>
#include <boost/tokenizer.hpp>

#include <cerrno>
#include <cstring>

extern "C" {
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
}

using namespace Hypertable;
using namespace std;


void
ClusterDefinition::ClusterDefinition::get_role_members(const string &role,
                                                       vector<string> &members,
                                                       int64_t *generation){
  lock_guard<mutex> lock(m_mutex);
  load_file();
  auto iter = m_roles.find(role);
  if (iter == m_roles.end()) {
    members.clear();
    if (generation)
      *generation = 0;
    return;
  }
  members = iter->second;
  if (generation)
    *generation = (int64_t)m_last_mtime;
}

void ClusterDefinition::ClusterDefinition::load_file() {
  struct stat stats;

  if (stat(m_fname.c_str(), &stats)) {
    if (errno == ENOENT) {
      m_roles.clear();
      m_last_mtime = 0;
      return;
    }
    HT_FATALF("Unable to access file %s - %s",
              m_fname.c_str(), strerror(errno));
  }

  if (stats.st_mtime > m_last_mtime) {
    TokenizerPtr tokenizer(new Tokenizer(m_fname));
    Token token;
    while (tokenizer->next(token)) {
      if (token.type == Token::ROLE)
        add_role(token.text);
    }
    m_last_mtime = stats.st_mtime;
  }
  
}

void ClusterDefinition::ClusterDefinition::add_role(const std::string &text) {
  const char *base = text.c_str();
  const char *ptr;
  for (ptr=base; *ptr && !isspace(*ptr); ptr++)
    ;
  string token(base, ptr-base);
  boost::trim(token);
  if (strcasecmp(token.c_str(), "role:"))
    HT_THROWF(Error::SYNTAX_ERROR, "Mal-formed role definition: %s", text.c_str());
  while (*ptr && isspace(*ptr))
    ptr++;
  base = ptr;
  while (*ptr && !isspace(*ptr))
    ptr++;
  string name(base, ptr-base);
  boost::trim(name);
  if (name.empty())
    HT_THROWF(Error::SYNTAX_ERROR, "Mal-formed role definition: %s", text.c_str());
  string input(ptr);
  string value;

  boost::char_separator<char> sep(" \t\n\r");
  boost::tokenizer< boost::char_separator<char> > tokens(input, sep);
  for (auto &t : tokens)
    value += t + " ";
  boost::trim(value);

  m_roles[name] = HostSpecification(value);
}
