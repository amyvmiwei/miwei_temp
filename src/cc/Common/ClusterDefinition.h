/* -*- c++ -*-
 * Copyright (C) 2007-2015 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 3 of the
 * License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

/// @file
/// Declarations for ClusterDefinition.
/// This file contains type declarations for ClusterDefinition, a class that
/// caches and provides access to the information in a cluster definition file.

#ifndef Common_ClusterDefinition_ClusterDefinition_h
#define Common_ClusterDefinition_ClusterDefinition_h

#include <ctime>
#include <mutex>
#include <string>
#include <unordered_map>
#include <vector>

namespace Hypertable {
namespace ClusterDefinition {

  /// @addtogroup Common
  /// @{

  /// Represents a cluster definition.
  class ClusterDefinition {

  public:

    /// Constructor.
    /// @param fname Pathname of cluster definition file
    ClusterDefinition(const std::string &fname) : m_fname(fname) {};

    /// Get list of members of a role
    /// This method first refreshes its view of the cluster definition by
    /// reloading the cluster definition file, #m_fname, if it has not been
    /// modified since the last refresh.  If a role by the name of
    /// <code>role</code> exists in the cluster definition, <code>members</code>
    /// is populated with the corresponding members.  If the
    /// <code>generation</code> parameter is not nullptr, then it is set to the
    /// generation number of the cluster definition file (last modification
    /// time).  If the cluster definition file, #m_fname, does not exist,
    /// <code>members</code> is cleared and the <code>generation</code> is set
    /// to 0.
    /// @param role Role name
    /// @param members Reference to vector to hold members
    /// @param generation Address of variable populated with definition generation number
    void get_role_members(const std::string &role,
                          std::vector<std::string> &members,
                          int64_t *generation = nullptr);

  private:

    /// Load cluster definition from file
    /// Loads the cluster definition from the file #m_fname, if the file was
    /// modified since the last time it was loaded.
    void load_file();

    /// Parses a role definition statement
    /// Parses the role definition statement contained in <code>text</code> and
    /// adds the information to #m_roles.
    /// @param text
    void add_role(const std::string &text);

    /// Mutex for serializing access to members
    std::mutex m_mutex;

    /// Name of cluster definition file
    std::string m_fname;

    /// Last modification time of cluster definition file
    time_t m_last_mtime {};

    /// Map from role names to members
    std::unordered_map<std::string, std::vector<std::string>> m_roles;

  };

  /// @}

}}

#endif // Common_ClusterDefinition_ClusterDefinition_h
