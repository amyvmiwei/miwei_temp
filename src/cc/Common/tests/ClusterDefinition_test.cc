
#include <Common/Compat.h>
#include <Common/Error.h>
#include <Common/FileUtils.h>
#include <Common/Logger.h>
#include <Common/ClusterDefinition.h>

#include <chrono>
#include <fstream>
#include <iostream>
#include <string>
#include <thread>
#include <vector>

using namespace Hypertable;
using namespace std;

namespace {

  const char *examples[] = {
R"(INSTALL_PREFIX=/opt/hypertable
HYPERTABLE_VERSION=0.9.8.9

role: source test00
role: master test[00-02]
role: hyperspace test[00-02]
role: spare

include: 'core.tasks')",
R"(INSTALL_PREFIX=/opt/hypertable
HYPERTABLE_VERSION=0.9.8.9

role: source test00
role: master test[00-05]
- test03
role: hyperspace test[00-02]
role: spare

include: 'core.tasks')",
R"(INSTALL_PREFIX=/opt/hypertable
HYPERTABLE_VERSION=0.9.8.9

role: source test00
role: master test[00-05] - test04

include: 'core.tasks')",
    nullptr
  };

  void display_members(ofstream &out, const string &role, vector<string> &members) {
    bool first {true};
    out << role << ": ";
    for (const auto &member : members) {
      if (!first)
        out << ", ";
      else
        first = false;
      out << member;
    }
    out << endl;
  }

}

int main(int argc, char **argv) {
  unique_ptr<ClusterDefinition::ClusterDefinition> cluster_def(new ClusterDefinition::ClusterDefinition("cluster.def"));
  vector<string> members;
  int64_t generation {};
  int64_t previous_generation {};

  {
    ofstream out("ClusterDefinition_test.output");

    for (int i=0; examples[i]; i++) {
      {
        ofstream cluster_file_out("cluster.def");
        cluster_file_out << examples[i];
      }
      cluster_def->get_role_members("master", members, &generation);
      HT_ASSERT(generation != previous_generation);
      previous_generation = generation;
      display_members(out, "master", members);
      this_thread::sleep_for(chrono::milliseconds(2000));
    }

    cluster_def->get_role_members("master", members);
    display_members(out, "master", members);

    FileUtils::unlink("cluster.def");

    cluster_def->get_role_members("master", members, &generation);
    HT_ASSERT(generation == 0);
    previous_generation = generation;
    display_members(out, "master", members);

    {
      ofstream cluster_file_out("cluster.def");
      cluster_file_out << examples[0];
    }

    cluster_def->get_role_members("master", members, &generation);
    display_members(out, "master", members);

    FileUtils::unlink("cluster.def");
  }

  string cmd = "diff ClusterDefinition_test.output ClusterDefinition_test.golden";
  if (system(cmd.c_str()) != 0)
    quick_exit(EXIT_FAILURE);

  quick_exit(EXIT_SUCCESS);
}
