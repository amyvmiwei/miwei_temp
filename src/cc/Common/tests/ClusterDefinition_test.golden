master: test00, test01, test02
master: test00, test01, test02, test04, test05
master: test00, test01, test02, test03, test05
master: test00, test01, test02, test03, test05
master: 
master: test00, test01, test02
